### ISSUE TITLE
std::vector garbage collection allows use after free

### ISSUE THREAD

**Issue by jakecobb** [swig/swig#945]
I noticed this in Python, haven't test other languages yet.  The `std::vector` support (from `std_vector.i`) for value (non-pointer) types allows the vector to be garbage collected while the managed layer still holds a pointer to one of its values.  Here is a SWIG example:

```
// test.i
%module test
%include "std_vector.i"
%include "std_string.i"

%inline %{
#include <vector>
#include <string>

struct Value {
  std::string name;
  Value(const std::string& name) : name(name) { }	
};

std::vector<Value> someValues() {
  std::vector<Value> values;
  values.emplace_back("test1");
  return values;
}
%}

%template(Values) std::vector<Value>;

```

In Python, you get some behavior like:
```python
>>> import test
>>> vals = test.someValues()
>>> v0 = vals[0]
>>> v0.name
'test1'
>>> del vals
>>> v0.name
''
```

Of course in my real use case the value has a `std::unique_ptr` leading to a crash on use after free.  An even shorter example of the problem when the `std::vector` is a temporary:

```python
>>> v = test.someValues()[0]
>>> v.name
''
```

It seems that either some back-reference is necessary to keep the vector from being freed while proxies to its values are alive or the index operator should be making a copy that the managed layer owns.

**Comment by jakecobb** [comment]
Here is a possible workaround:

```
%fragment("container_owner_attribute", "header") {
  PyObject* container_owner_attribute() {
    static PyObject* name = SWIG_Python_str_FromChar("__owner");
    return name;
  }
}

%define %StdVector(_name, _type)
// override the default SWIGTYPE out typemap
%typemap(out, noblock=1, fragment="container_owner_attribute") _type *, _type &, _type &&, _type[] {
  %set_output(SWIG_NewPointerObj(%as_voidptr($1), $descriptor, $owner | %newpointer_flags));
  // set a reference to the owner
  PyObject_SetAttr($result, container_owner_attribute(), $self);
}
// now declare the template with the above typemap operating
%template(_name) std::vector<_type>;
// remove the typemap to revert to the default typemap
%clear _type *, _type &, _type &&, _type[];
%enddef

%StdVector(Values, Value);
```

Of course, this only helps for Python and assumes there is no pre-existing typemap on `_type` that needs to be preserved.  It does prevent the use-after-free unless the user explicitly deletes `__owner` from the `std::vector` wrapper object.

Also, since this behavior seems to come from `pycontainer.swg`, I expect all standard container classes to be affected and not just `std::vector`.

**Comment by wsfulton** [comment]
Fixed by #1234 for Python anyway. Separate pull requests along the same lines for other target languages are welcome.
